File: algorithms/acwing/binarysearch/BinarySearch.java
package blogandquestion.algorithms.acwing.binarysearch;

/**
 * @author Heming
 * created by IntelliJ IDEA
 * founded at 2021/06/05 09:43:12
 * description:
 */
public class BinarySearch {
    // 找到大于等于tar的最后一个位置（如果数组中存在tar，返回tar出现的最后位置，否则返回大于tar的第一个位置）
    public int find(int[] arr, int tar) {
        int l = 0, r = arr.length - 1;
        while (l < r) {
            int mid = (l + r + 1) >>> 1;
            if (arr[mid] <= tar) {
                l = mid;
            } else {
                r = mid - 1;
            }
        }
        return l;
    }

    // upper_bound返回第一个大于x的位置
    public int upper_bound(int[] arr, int x) {
        int l = 0, r = arr.length - 1;
        while (l < r) {
            int mid = (l + r) >>> 1;
            if (arr[mid] > x) {
                r = mid;
            } else {
                l = mid + 1;
            }
        }
        return l;
    }

    // lower_bound函数返回第一个大于等于x的位置
    public int lower_bound(int[] arr, int x) {
        int l = 0, r = arr.length - 1;
        while (l < r) {
            int mid = (l + r) >>> 1;
            if (arr[mid] >= x) {
                r = mid;
            } else {
                l = mid + 1;
            }
        }
        return l;
    }

    public static void main(String[] args) {
        int[] test = {1, 2, 5, 5, 6, 7, 7, 9};
        BinarySearch bs = new BinarySearch();
        int res = bs.lower_bound(test, 6);
        System.out.println(res);
    }
}

File: src/binarysearch/binarySearch.js

function binarySearch(arr, target) {
  let left = 0, right = arr.length;
  while (left < right) {
    let mid = (left + right) >>> 1;
    let t = arr[mid];
    if (t == target) {
      return mid;
    } else if (t > target) {
      right = mid;
    } else {
      left = mid + 1;
    }
  }
  return -1;
}

let arr = [1,2,2,3,3,3,4,6];
console.log(binarySearch(arr, 3));
console.log(lower_bound(arr, 5));
console.log(upper_bound(arr,5));

function lower_bound(arr, target) {
  let left = 0, right = arr.length;
  while (left < right) {
    let mid = (left + right) >>> 1;
    if (arr[mid] >= target) {
      right = mid;
    } else {
      left = mid + 1;
    }
  }
  return left;
}

function upper_bound(arr, target) {
  let left = 0, right = arr.length;
  while (left < right) {
    let mid = (left + right) >>> 1;
    if (arr[mid] > target) {
      right = mid;
    } else {
      left = mid + 1;
    }
  }
  return left;
}


console.log(lower_bound([1, 2, 3, 5], 6));
